aspy/refactor_imports/classify.py

```diff
@@ -110,7 +110,10 @@ def _get_module_info(module_name, application_dirs):
         elif spec.origin in {None, 'namespace'}:
             return True, next(iter(spec.submodule_search_locations)), False
         # special case pypy3 bug(?)
-        elif not os.path.exists(spec.origin):
+        elif (
+                not os.path.exists(spec.origin)
+                and '.zip' not in spec.origin.lower()
```

Comment:
this seems like a very weak condition, any way to improve on this?

Reply:
I thought about authoring a new function that walks up the directory tree checking each level for being a zip file (or a simpler check for having a `.zip` extension).  I tried to balance that against your request for a "small patch".

Happy to switch to either form of the directory walker if you prefer.  Not sure what the performance cost of `zipfile.is_zipfile` is if we go that route (or how much you're valuing performance here...)  It's just checking a magic number, but that does require file I/O.

Reply:
what does the `.origin` look like here?

Reply:
`'C:\\Users\\shea.parkes\\AppData\\Local\\Temp\\pytest-of-Shea.Parkes\\pytest-33\\test_classify_pythonpath_zipim0\\ppth\\fzip.zip\\fzip.py'`

Reply:
hmmm while we're on the topic of edge cases, I did a little digging -- it seems the standard library can also be distributed in this same way (as zipfiles) (this is why, for instance, the standard `sys.path` looks like this):

```pycon
>>> pprint.pprint(sys.path)
['',
 '/usr/lib/python36.zip',
 '/usr/lib/python3.6',
 '/usr/lib/python3.6/lib-dynload',
 '/usr/local/lib/python3.6/dist-packages',
 '/usr/lib/python3/dist-packages']
```

this is commonly done in embedding scenarios, and especially on windows

Reply:
a stronger assertion would be `'.zip{}'.format(os.sep)'` in ...` -- or maybe utilize that the `loader` is a special class

Reply:
Yea, the full spec is:
```
ModuleSpec(name='fzip', loader=<zipimporter object "C:\Users\shea.parkes\AppData\Local\Temp\pytest-of-Shea.Parkes\pytest-33\test_classify_pythonpath_zipim0\ppth\fzip.zip">, origin='C:\\Users\\shea.parkes\\AppData\\Local\\Temp\\pytest-of-Shea.Parkes\\pytest-33\\test_classify_pythonpath_zipim0\\ppth\\fzip.zip\\fzip.py')
```

And:
```
(Pdb) p type(spec.loader)
<class 'zipimport.zipimporter'>
```

Could do an `not isinstance(...)` on the loader.  Would you like me to try that?

I would hope the actual standard libraries would be caught in the earlier logic gates.

Reply:
> I would hope the actual standard libraries would be caught in the earlier logic gates.

this one is pretty early in the chain -- I'm not sure it would :/ -- since you're on windows can you download one of the "embeddable" installs from python.org and see what you get for

```python
import importlib.util
importlib.util.find_spec('email')
```

Reply:
I went ahead and switched it to do `not isinstance(spec.loader, zipimport.zipimporter)`.  Downloading the embedded python to see how that behaves now.

Reply:
Here's the spec for `email` in an embeddable world:
```
C:\DevApps\python-3.8.1-embed-amd64>python
Python 3.8.1 (tags/v3.8.1:1b293b6, Dec 18 2019, 23:11:46) [MSC v.1916 64 bit (AMD64)] on win32
>>> import sys
>>> sys.executable
'C:\\DevApps\\python-3.8.1-embed-amd64\\python.exe'
>>> import importlib.util
>>> spec_email = importlib.util.find_spec('email')
>>> spec_email
ModuleSpec(name='email', loader=<zipimporter object "C:\DevApps\python-3.8.1-embed-amd64\python38.zip\">, origin='C:\\DevApps\\python-3.8.1-embed-amd64\\python38.zip\\email\\__init__.pyc', submodule_search_locations=['C:\\DevApps\\python-3.8.1-embed-amd64\\python38.zip\\email'])
>>>
```

My battery's running low, but I'll see if I can follow the logic gates to confirm whether it would work (and/or see about adding a test to try and replicate it).

Reply:
Hrm, in the embedded scenario:
```
>>> sys.builtin_module_names
('_abc', '_ast', '_bisect', '_blake2', '_codecs', '_codecs_cn', '_codecs_hk', '_codecs_iso2022', '_codecs_jp', '_codecs_kr', '_codecs_tw', '_collections', '_contextvars', '_csv', '_datetime', '_functools', '_heapq', '_imp', '_io', '_json', '_locale', '_lsprof', '_md5', '_multibytecodec', '_opcode', '_operator', '_pickle', '_random', '_sha1', '_sha256', '_sha3', '_sha512', '_signal', '_sre', '_stat', '_statistics', '_string', '_struct', '_symtable', '_thread', '_tracemalloc', '_warnings', '_weakref', '_winapi', '_xxsubinterpreters', 'array', 'atexit', 'audioop', 'binascii', 'builtins', 'cmath', 'errno', 'faulthandler', 'gc', 'itertools', 'marshal', 'math', 'mmap', 'msvcrt', 'nt', 'parser', 'sys', 'time', 'winreg', 'xxsubtype', 'zlib')
```

Could perhaps do a check based on the fact that `spec_email.origin` is in the same directory as `sys.executable`:
```
>>> spec_email.origin
'C:\\DevApps\\python-3.8.1-embed-amd64\\python38.zip\\email\\__init__.pyc'
>>> sys.executable
'C:\\DevApps\\python-3.8.1-embed-amd64\\python.exe'
>>>
```

Reply:
`classify.py` has no external dependencies, might be able to `python -i classify.py` and see what it thinks about things before/after

Reply:
Have to shut down for battery (and sleep).  I'd welcome any suggestions (or commits if you want to take this over yet tonight).  I'll try to give some thought as to how well the prior logic gates would have supported embedded versions as well.

Reply:
I went ahead and put in a logic gate based on the location of `sys.executable` and `spec.loader.archive`.  I pushed that one up already.

Here's some embedded REPL for support of that:
```
Python 3.8.1 (tags/v3.8.1:1b293b6, Dec 18 2019, 23:11:46) [MSC v.1916 64 bit (AMD64)] on win32
>>> import sys
>>> import os.path
>>> import importlib.util
>>> email_spec = importlib.util.find_spec('email')
>>> os.path.dirname(sys.executable)
'C:\\DevApps\\python-3.8.1-embed-amd64'
>>> os.path.dirname(email_spec.loader.archive)
'C:\\DevApps\\python-3.8.1-embed-amd64'
>>> os.path.samefile(os.path.dirname(sys.executable), os.path.dirname(email_spec.loader.archive))
True
>>>
```

I suppose you could mock `sys.executable` in some manner to test that.  Let me know if you want to try and go there.

I'll spend the last few minutes of the commute seeing about `python -i classify.py` with the embedded `python.exe`.

Reply:
Alright, got the `-i classify.py` working enough to debug `classify_import('email')` in an embedded world.

So even without the new check, `classify_import('email')` was returning `BUILTIN` due to this conditional in `def classify_import`:
```
    # Otherwise we assume it is a system module or a third party module
    elif (
            found and
            PACKAGES_PATH not in module_path and
            not _due_to_pythonpath(module_path)
    ):
        return ImportType.BUILTIN
```

I think it's still reasonable to add the conditional I did to have a more explicit catch for embedded.  But let me know if you'd rather remove it.

Also let me know if you'd like me to do some kind of mocking to try and test that.  I'm guessing you could get `sys.executable` to return what you want it to for a given test.

I'm going to step away for now and await your feedback.

Thanks.

Reply:
Curiousity got the better of me, and I think I figured out how to patch a module attribute.  I went ahead and pushed up a test case that attempts to cover the embedded scenario.  I'm going to go ahead and mark this particular github comment chain as resolved, but you are of course welcome to re-open or start any additional comments/requests.